### PR TITLE
Handle native backend tensors in data adapters

### PR DESCRIPTION
**Pull request by samudraneel05**
as suggested by @hertschuh in keras-team/keras-mlx#40

## Description

the data adapters should handle native arrays where `is_tensor(x)` is true, rather than relying on a keras-mlx workaround that converts the full dataset to numpy.

Pluggable backend tensors (e.g. MLX arrays) are rejected by the array data adapter. `can_slice_array()` and `convert_to_sliceable()` check for numpy, TF, JAX, torch, pandas, scipy, and `__array__`, but MLX intentionally dropped `__array__` in favor of the Python Buffer Protocol (ml-explore/mlx#323). So MLX arrays fall through every check and `model.fit/evaluate/predict` crashes.

## Fix

added `ops.is_tensor(x)` as a fallback in both `can_slice_array()` and `convert_to_sliceable()`. For the sliceable representation, using `np.asarray(x)` to get a zero-copy numpy view. falling back to `ops.convert_to_numpy(x)` when `np.asarray` raises (e.g. MLX bfloat16, which fails the buffer protocol) or silently wraps the input in a 0-D object array (e.g. openvino tensors, which don't expose `__array__` or the buffer protocol).

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/trainers/data_adapters/array_slicing.py [hunk 1]
```diff
@@ -333,6 +334,7 @@ def can_slice_array(x):
         or data_adapter_utils.is_pandas_data_frame(x)
         or data_adapter_utils.is_pandas_series(x)
         or hasattr(x, "__array__")
+        or ops.is_tensor(x)
```
[the remark this answers is not in the document]
Comment (samudraneel05) [reply keras/src/trainers/data_adapters/array_slicing.py]: Noted, fair enough

File: keras/src/trainers/data_adapters/array_slicing.py [hunk 2]
```diff
@@ -395,6 +396,16 @@ def convert_single_array(x):
...
+        elif backend.ops.is_tensor(x):
+            # Generic fallback for pluggable backend tensors (e.g. MLX).
+            try:
+                res = np.asarray(x)
+                if res.dtype == object:
+                    raise ValueError("np.asarray returned an object array")
+                x = res
+            except (ValueError, TypeError, RuntimeError):
+                x = backend.ops.convert_to_numpy(x)
+            sliceable_class = NumpySliceable
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/array_slicing.py]: Actually, I had something different in mind. The whole point being that the NumPy intermediary is not necessary.

There would be a new Sliceable subclass called `NativeArraySliceable` with the following class overrides:

- `cast`: uses `backend.cast`
- `convert_to_numpy`: uses `backend.convert_to_numpy`
- `convert_to_tf_dataset_compatible` should `raise NotImplementedError` because it should never be used
- `convert_to_jax_compatible` should `raise NotImplementedError` because it should never be used
- `convert_to_torch_compatible` should `raise NotImplementedError` because it should never be used

Then, every `DataAdapter` subclass should have a `get_native_iterator` function. By default, it should call `get_numpy_iterator`. But for `ArrayDataAdapter`, it should use the `NativeArraySliceable`.

Then `MLXEpochIterator` should use `get_native_iterator` instead of `get_numpy_iterator`.

File: keras/src/trainers/data_adapters/array_slicing.py [hunk 3]
```diff
@@ -395,6 +396,16 @@ def convert_single_array(x):
         elif hasattr(x, "__array__"):
             x = np.asarray(x)
             sliceable_class = NumpySliceable
+        elif backend.ops.is_tensor(x):
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/array_slicing.py]: The `elif backend.ops.is_tensor(x):` should come before the `elif hasattr(x, "__array__"):`

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for working on this!

CC @MarcosAsh

File: keras/src/trainers/data_adapters/array_slicing.py (hunk 2 above)
Comment (samudraneel05) [reply keras/src/trainers/data_adapters/array_slicing.py]: ahh okay, thank you for being so explicit & descriptive. this helps a lot.

File: keras/src/trainers/data_adapters/array_slicing.py (hunk 3 above)
Comment (samudraneel05) [reply keras/src/trainers/data_adapters/array_slicing.py]: noted, thank you.

**Comment by samudraneel05** [comment]
i added three tests.

one checks that adapters without a native iterator fall back to the numpy iterator, one checks that native arrays remain native across regular and shuffled batching, and one checks the conversion methods on nativearraysliceable. the small fake array is there so the generic pluggable backend path can be tested in keras without depending on mlx or another backend package.

once this goes through, the tests on keras-team/keras-mlx#57 should turn green too.

also added the `target backend in (None, "numpy"):` condition so native sliceables are only used by the native and numpy iterator paths (jax tests were failing otherwise due to nativearraysliceable being selected by the jax iterator)

File: keras/src/trainers/data_adapters/array_slicing.py [hunk 4]
```diff
@@ -109,6 +109,33 @@ class NumpySliceable(Sliceable):
...
+    def convert_to_tf_dataset_compatible(cls, x):
+        raise NotImplementedError
+
+    @classmethod
+    def convert_to_jax_compatible(cls, x):
+        raise NotImplementedError
+
+    @classmethod
+    def convert_to_torch_compatible(cls, x):
+        raise NotImplementedError
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/array_slicing.py]: You did capture exactly what I described, thanks!

Now, after looking at it and looking at the unit tests, I changed my mind a little bit. Basically, I'd like to have a real example of `NativeArraySliceable` being used within the keras repo. And that would be with the torch backend.

So can you:
- remove `data_adapter_utils.is_torch_tensor`
- remove `TorchSliceable`
- rename all `convert_to_torch_compatible` methods to `convert_to_native_compatible` in all Sliceable classes

Then here in `NativeArraySliceable`, remove the `convert_to_tf_dataset_compatible`, `convert_to_jax_compatible` and `convert_to_torch_compatible` overrides, let's keep the default implementation of the super class.

That means that with the torch backend, torch tensors will now use `NativeArraySliceable`. I think it should work with all the current tests.

File: keras/src/trainers/data_adapters/array_slicing.py [hunk 5]
```diff
@@ -392,6 +420,8 @@ def convert_single_array(x):
             sliceable_class = PandasSeriesSliceable
         elif data_adapter_utils.is_scipy_sparse(x):
             sliceable_class = ScipySparseSliceable
+        elif backend.ops.is_tensor(x) and target_backend in (None, "numpy"):
+            sliceable_class = NativeArraySliceable
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/array_slicing.py]: Why `target_backend in (None, "numpy")`? I think we should remove that condition.

And why the "numpy" part?

File: keras/src/trainers/data_adapters/array_data_adapter_test.py [hunk 6]
```diff
@@ -12,9 +14,91 @@
...
+
+    def __array__(self, dtype=None, copy=None):
+        raise AssertionError(
+            "Native arrays should not be converted through NumPy"
+        )
+
+    def __getitem__(self, indices):
+        if isinstance(indices, np.ndarray):
+            raise ValueError("NumPy indices are unsupported")
+        return _NativeArray(self.array[indices])
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/array_data_adapter_test.py]: Remove this, let's use real arrays (see below).

File: keras/src/trainers/data_adapters/array_data_adapter_test.py [hunk 7]
```diff
@@ -12,9 +14,91 @@
...
 
 
 class TestArrayDataAdapter(testing.TestCase):
+    def test_native_iterator_default(self):
+        class TestAdapter(DataAdapter):
+            def get_numpy_iterator(self):
+                return iter([np.array([1, 2])])
+
+        batch = next(TestAdapter().get_native_iterator())
+        self.assertAllEqual(batch, [1, 2])
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/array_data_adapter_test.py]: I'm not sure we need this test when we're already testing the real `ArrayDataAdapter`.

File: keras/src/trainers/data_adapters/array_data_adapter_test.py [hunk 8]
```diff
@@ -12,9 +14,91 @@
...
+
+        cast.assert_called_once_with(x, "float16")
+        convert_to_numpy.assert_called_once_with(x)
+        for method in (
+            array_slicing.NativeArraySliceable.convert_to_tf_dataset_compatible,
+            array_slicing.NativeArraySliceable.convert_to_jax_compatible,
+            array_slicing.NativeArraySliceable.convert_to_torch_compatible,
+        ):
+            with self.assertRaises(NotImplementedError):
+                method(x)
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/array_data_adapter_test.py]: Let's not do that. Let's test in real conditions, which means:
- remove the mocking
- just do `x = backend.ops.reshape(backend.ops.arange(24, dtype="float32"), (8, 3))`

But that's already covered in the tests below, so I don't think we need these at all. In particular, `NativeArraySliceable` should be covered with the torch backend, which we can verify in the code coverage.

So remove these tests.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Did you get a chance to test that it works with the MLX backend?

File: keras/src/trainers/data_adapters/array_slicing.py (hunk 5 above)
Comment (samudraneel05) [reply keras/src/trainers/data_adapters/array_slicing.py]: the current JAX class name changed from `Array` to `jaxlib._jax.Array`, so Keras’s detector was missing it. thereforce i had added the secondary condition to prevent getting NotImplementedError due to the fallback path getting selected. with the latest changes that i'm thinking of, this should not be needed as i'll just fix the root cause at `is_jax_array():` definition

File: keras/src/trainers/data_adapters/array_data_adapter_test.py (hunk 6 above)
Comment (samudraneel05) [reply keras/src/trainers/data_adapters/array_data_adapter_test.py]: Understood, removing all these mock tests

File: keras/src/trainers/data_adapters/array_slicing.py (hunk 4 above)
Comment (samudraneel05) [reply keras/src/trainers/data_adapters/array_slicing.py]: i've tried to incorporate these points in my changes below.

**Comment by samudraneel05** [comment]
removed the mock tests and made the torch related changes. previously for the NativeArraySliceable I was using `convert_to_tensor()`, which works perfectly fine with MLX but crashes on torch because it put indices on the active device (mps) while the input stayed on cpu.

since torch tensors will now use `NativeArraySliceable`i've now gone for the `indices.tolist()` approach as that works for both without any significant slowdown. 

i hope this approach makes sense.

**Comment by MarcosAsh** [comment]
Ran it against keras-mlx#57 on Apple silicon with the CI config, mlx 0.31.2: full keras suite 12938 passed, 0 failed. Trainers and callbacks also pass with compile on. Works for MLX.

**Review by hertschuh — APPROVED** [review]
Thanks!

**Comment by hertschuh** [comment]
@samudraneel05 

Thanks for this PR!

Would you be able to re-apply this to the `master` branch? Thanks!

**Comment by samudraneel05** [comment]
Thank you for the help and the reviews, and yes, on it, will raise a PR and tag you and Marcos once done. In the meantime, keras-team/keras-mlx#57 can now be merged I believe
